core/undo_redo.cpp

```diff
@@ -268,7 +268,24 @@ void UndoRedo::_process_operation_list(List<Operation>::Element *E) {
 
 			case Operation::TYPE_METHOD: {
 
-				obj->call(op.name, VARIANT_ARGS_FROM_ARRAY(op.args));
+				Vector<const Variant *> argptrs;
+				argptrs.resize(VARIANT_ARG_MAX);
+				int argc = 0;
+
+				for (int i = 0; i < VARIANT_ARG_MAX; i++) {
+					if (op.args[i].get_type() == Variant::NIL) {
+						break;
+					}
+					argptrs.write[i] = &op.args[i];
+					argc++;
+				}
+				argptrs.resize(argc);
+
+				Variant::CallError ce;
+				obj->call(op.name, (const Variant **)argptrs.ptr(), argc, ce);
```

Comment:
You could probably reuse `(const Variant **)argptrs.ptr()` in `method_callback` and completely drop `VARIANT_ARGS_FROM_ARRAY` from `core/object.h`, those two were the only uses.

Why was it necessary to reimplement this manually instead of using `VARIANT_ARGS_FROM_ARRAY` though? Is it correct to skip `Variant::NIL` arguments?

Reply:
I don't see another way to do it.
Using `VARIANT_ARGS_FROM_ARRAY` creates multiple values this is what `call` expects.
https://github.com/godotengine/godot/blob/f5a8637e7aa6561db811548b419839b5bfa32fd4/core/object.cpp#L879

But a call with error report `ce` expects one `Variant`
https://github.com/godotengine/godot/blob/f5a8637e7aa6561db811548b419839b5bfa32fd4/core/object.cpp#L911

It stops at the first `Variant::NIL` because `null` is not working at all.
`op.args` always contains 5 elements(filled with Variant::NIL), even if less parameters are provided in `undo.add_undo_method(...)`.
This is still an issue if you want to set something to `null` since `null == Variant::NIL`
https://github.com/godotengine/godot/blob/f5a8637e7aa6561db811548b419839b5bfa32fd4/core/object.h#L45

I had to reimplement it to get the right amount of valid parameters `argc`

Is there a way to get the amount of parameters provided in `VARIANT_ARGPTRS` ? Maybe this would solve the `null` issue.
https://github.com/godotengine/godot/blob/f5a8637e7aa6561db811548b419839b5bfa32fd4/core/undo_redo.cpp#L130
https://github.com/godotengine/godot/blob/f5a8637e7aa6561db811548b419839b5bfa32fd4/core/undo_redo.cpp#L147

Reply:
Ah indeed, thanks for the detailed explanation. Let's merge as is and see how it goes then :)